pkg/reconciler/tenancy/initialization/apibinder_initializer_reconcile.go

```diff
@@ -0,0 +1,290 @@
+					logger.V(2).Info("APIBinding already exists")
+					continue
+				}
+
+				errors = append(errors, err)
+				continue
+			}
+
+			logger.V(2).Info("created APIBinding")
+		}
+	}
+
+	if len(errors) > 0 {
+		logger.Error(utilerrors.NewAggregate(errors), "error initializing APIBindings")
+
+		conditions.MarkFalse(
+			clusterWorkspace,
+			WorkspaceAPIBindingsInitialized,
+			WorkspaceInitializedAPIBindingErrors,
+			conditionsv1alpha1.ConditionSeverityError,
+			"encountered errors: %v",
+			utilerrors.NewAggregate(errors),
+		)
+
+		if someExportsMissing {
+			// Retry if any APIExports are missing, as it's possible they'll show up (cache server slow to catch up,
+			// arrive via replication)
+			return utilerrors.NewAggregate(errors)
+		}
+
+		return nil
+	}
+
+	// Make sure all the expected bindings are there & ready to use
+	var incomplete []string
+
+	for exportRef := range requiredExportRefs {
+		workspaceExportRef := apisv1alpha1.WorkspaceExportReference{
+			Path:       exportRef.Path,
+			ExportName: exportRef.Name,
+		}
+
+		binding, exists := exportToBinding[workspaceExportRef]
+		if !exists {
+			incomplete = append(incomplete, fmt.Sprintf("for APIExport %s|%s", exportRef.Path, exportRef.Name))
+			continue
+		}
+
+		if !conditions.IsTrue(binding, apisv1alpha1.InitialBindingCompleted) {
+			incomplete = append(incomplete, binding.Name)
+		}
+	}
+
+	if len(incomplete) > 0 {
+		sort.Strings(incomplete)
+
+		conditions.MarkFalse(
+			clusterWorkspace,
+			WorkspaceAPIBindingsInitialized,
+			WorkspaceInitializedWaitingOnAPIBindings,
+			conditionsv1alpha1.ConditionSeverityInfo,
+			"APIBinding(s) not yet fully bound: %s", strings.Join(incomplete, ", "),
+		)
+
+		return nil
+	}
+
+	clusterWorkspace.Status.Initializers = initialization.EnsureInitializerAbsent(tenancyv1alpha1.ClusterWorkspaceAPIBindingsInitializer, clusterWorkspace.Status.Initializers)
+
+	return nil
+}
+
+// maxExportNamePrefixLength is the maximum allowed length for the export name portion of the generated API binding
+// name. Subtrace 1 for the dash ("-") that separates the export name prefix from the hash suffix, and 5 for the
+// hash length.
+const maxExportNamePrefixLength = validation.DNS1123SubdomainMaxLength - 1 - 5
+
+func generateAPIBindingName(clusterName logicalcluster.Name, exportPath, exportName string) string {
+	maxLen := len(exportName)
+	if maxLen > maxExportNamePrefixLength {
+		maxLen = maxExportNamePrefixLength
+	}
+
+	exportNamePrefix := exportName[:maxLen]
+
+	hash := toBase36Sha224(
+		clusterName.String() + "|" + exportPath + "|" + exportName,
+	)
+
+	hash = hash[0:5]
+	// Have to lowercase because Kubernetes names must be lowercase
+	hash = strings.ToLower(hash)
+
+	return fmt.Sprintf("%s-%s", exportNamePrefix, hash)
+}
+
+func toBase36(hash [28]byte) string {
+	var i big.Int
+	i.SetBytes(hash[:])
+	return i.Text(62)
+}
+
+func toBase36Sha224(s string) string {
+	return toBase36(sha256.Sum224([]byte(s)))
+}
```

Comment:
@stevekuznetsov another canidate for kcp-dev/utils.

Reply:
`kcp-dev/utils` as a whole project sounds very similar to what a `utils` pkg is. we should be careful what to put there and constraint the scope.